Black is ignoring pyproject.toml

## Information

**VIM version**

NVIM v0.3.1                                                                     
Build type: Release                                                             

Operating System: ubuntu 18.04

## What went wrong

black ignores any pyproject.toml hence does not respect the project specific settings.

## Reproducing the bug

1. Enable black as a fixer
2. Open a python file with a pyproject.toml next to it: pyproject.toml must not be in one of the directory of the current workdir path.
3. When saving, the options from pyproject.toml are not respected

## Possible fix:

Black should be runned with a current work dir set to the directory of the file opened in the buffer.

### :ALEInfo


 Current Filetype: python
Available Linters: ['flake8', 'mypy', 'prospector', 'pycodestyle', 'pydocstyle', 'pyflakes', 'pylint', 'pyls', 'pyre', 'vulture']
  Enabled Linters: ['flake8']
 Suggested Fixers: 
  'add_blank_lines_for_python_control_statements' - Add blank lines before control statements.
  'autopep8' - Fix PEP8 issues with autopep8.
  'black' - Fix PEP8 issues with black.
  'isort' - Sort Python imports with isort.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'yapf' - Fix Python files with yapf.
 Linter Variables:

let g:ale_python_flake8_auto_pipenv = 0
let g:ale_python_flake8_change_directory = 1
let g:ale_python_flake8_executable = 'flake8'
let g:ale_python_flake8_options = ''
let g:ale_python_flake8_use_global = 0
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 1
let g:ale_fixers = {'rust': ['trim_whitespace', 'rustfmt'], 'php': ['remove_trailing_lines', 'trim_whitespace', 'phpcbf'], '*': ['remove_trailing_lines', 'trim_whitespace'], 'javascript': ['prettier', 'eslint'], 'eml': ['elm-format'], 'python': ['isort', 'black']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 1000
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 0
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'never'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let b:ale_linters = ['flake8']
let g:ale_linters_explicit = 0
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 'on_save'
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) flake8
(finished - exit code 0) ['/bin/bash', '-c', '''flake8'' --version']
(finished - exit code 0) ['/bin/bash', '-c', 'cd ''xxx/tests'' && ''flake8'' --format=default --stdin-display-name ''xxx/tests/test_export.py'' - < ''/tmp/nvimnICa1C/2/test_export.py''']

<<<NO OUTPUT RETURNED>>>

(finished - exit code 0) ['/bin/bash', '-c', 'cd ''xxx/tests'' && ''isort'' - < ''/tmp/nvimnICa1C/3/test_export.py''']
(finished - exit code 0) ['/bin/bash', '-c', '''black'' - < ''/tmp/nvimnICa1C/4/test_export.py''']
(finished - exit code 1) ['/bin/bash', '-c', 'cd ''xxx/tests'' && ''flake8'' --format=default --stdin-display-name ''xxx/tests/test_export.py'' - < ''/tmp/nvimnICa1C/5/test_export.py''']

<<<OUTPUT STARTS>>>
xxx/tests/test_export.py:14:80: E501 line too long (80 > 79 characters)
xxx/tests/test_export.py:33:80: E501 line too long (83 > 79 characters)
xxx/tests/test_export.py:37:80: E501 line too long (85 > 79 characters)
xxx/tests/test_export.py:40:80: E501 line too long (85 > 79 characters)
xxx/tests/test_export.py:53:80: E501 line too long (84 > 79 characters)
xxx/tests/test_export.py:61:80: E501 line too long (84 > 79 characters)
xxx/tests/test_export.py:71:80: E501 line too long (82 > 79 characters)
xxx/tests/test_export.py:73:80: E501 line too long (82 > 79 characters)
xxx/tests/test_export.py:75:80: E501 line too long (85 > 79 characters)
xxx/tests/test_export.py:95:80: E501 line too long (83 > 79 characters)
xxx/tests/test_export.py:141:80: E501 line too long (84 > 79 characters)
xxx/tests/test_export.py:147:80: E501 line too long (86 > 79 characters)
xxx/tests/test_export.py:165:80: E501 line too long (85 > 79 characters)
xxx/tests/test_export.py:170:80: E501 line too long (88 > 79 characters)
xxx/tests/test_export.py:176:80: E501 line too long (85 > 79 characters)
<<<OUTPUT ENDS>>>

--- a/autoload/ale/fixers/black.vim
+++ b/autoload/ale/fixers/black.vim
@@ -5,6 +5,7 @@ call ale#Set('python_black_executable', 'black')
 call ale#Set('python_black_use_global', get(g:, 'ale_use_global_executables', 0))
 call ale#Set('python_black_options', '')
 call ale#Set('python_black_auto_pipenv', 0)
+call ale#Set('python_black_change_directory', 1)
 
 function! ale#fixers#black#GetExecutable(buffer) abort
     if (ale#Var(a:buffer, 'python_auto_pipenv') || ale#Var(a:buffer, 'python_black_auto_pipenv'))
@@ -16,6 +17,10 @@ function! ale#fixers#black#GetExecutable(buffer) abort
 endfunction
 
 function! ale#fixers#black#Fix(buffer) abort
+    let l:cd_string = ale#Var(a:buffer, 'python_black_change_directory')
+    \   ? ale#path#BufferCdString(a:buffer)
+    \   : ''
+
     let l:executable = ale#fixers#black#GetExecutable(a:buffer)
 
     let l:exec_args = l:executable =~? 'pipenv$'
@@ -25,7 +30,7 @@ function! ale#fixers#black#Fix(buffer) abort
     let l:options = ale#Var(a:buffer, 'python_black_options')
 
     return {
-    \   'command': ale#Escape(l:executable) . l:exec_args
+    \   'command': l:cd_string . ale#Escape(l:executable) . l:exec_args
     \       . (!empty(l:options) ? ' ' . l:options : '')
     \       . ' -',
     \}

--- a/doc/ale-python.txt
+++ b/doc/ale-python.txt
@@ -139,6 +139,16 @@ g:ale_python_black_auto_pipenv                  *g:ale_python_black_auto_pipenv*
   Detect whether the file is inside a pipenv, and set the executable to `pipenv`
   if true. This is overridden by a manually-set executable.
 
+g:ale_python_black_change_directory       *g:ale_python_black_change_directory*
+                                          *b:ale_python_black_change_directory*
+  Type: |Number|
+  Default: `1`
+
+  If set to `1`, ALE will switch to the directory the Python file being
+  checked with `black` is in before checking it. This helps `black` find
+  configuration files more easily. This option can be turned off if you want
+  to control the directory Python is executed from yourself.
+
 
 ===============================================================================
 flake8                                                      *ale-python-flake8*

--- a/test/fixers/test_black_fixer_callback.vader
+++ b/test/fixers/test_black_fixer_callback.vader
@@ -24,11 +24,12 @@ After:
 Execute(The black callback should return the correct default values):
   silent execute 'file ' . fnameescape(g:dir . '/python_paths/with_virtualenv/subdir/foo/bar.py')
   AssertEqual
-  \ {'command': ale#Escape(ale#path#Simplify(g:dir . '/python_paths/with_virtualenv/env/' . b:bin_dir . '/black')) . ' -'},
+  \ {'command': ale#path#BufferCdString(bufnr('')) . ale#Escape(ale#path#Simplify(g:dir . '/python_paths/with_virtualenv/env/' . b:bin_dir . '/black')) . ' -'},
   \ ale#fixers#black#Fix(bufnr(''))
 
 Execute(The black callback should include options):
   let g:ale_python_black_options = '--some-option'
+  let g:ale_python_black_change_directory = 0
 
   silent execute 'file ' . fnameescape(g:dir . '/python_paths/with_virtualenv/subdir/foo/bar.py')
   AssertEqual
@@ -37,8 +38,10 @@ Execute(The black callback should include options):
 
 Execute(Pipenv is detected when python_black_auto_pipenv is set):
   let g:ale_python_black_auto_pipenv = 1
+  let g:ale_python_black_change_directory = 0
+
   call ale#test#SetFilename('/testplugin/test/python_fixtures/pipenv/whatever.py')
 
-  AssertEqual 
+  AssertEqual
   \ {'command': ale#Escape('pipenv') . ' run black -'},
   \ ale#fixers#black#Fix(bufnr(''))

